src/coordinator/storage/local/storage.go

```diff
@@ -190,7 +329,119 @@ func newWriteRequest(writeRequestCommon *writeRequestCommon, timestamp time.Time
 	}
 }
 
-func (s *localStorage) Close() error {
-	s.namespace.Finalize()
-	return nil
+type multiFetchResult struct {
+	sync.Mutex
+	result           *storage.FetchResult
+	err              xerrors.MultiError
+	dedupeFirstAttrs storage.Attributes
+	dedupeMap        map[string]multiFetchResultSeries
+}
+
+type multiFetchResultSeries struct {
+	idx   int
+	attrs storage.Attributes
+}
+
+func (r *multiFetchResult) add(
+	attrs storage.Attributes,
+	result *storage.FetchResult,
+	err error,
+) {
+	r.Lock()
+	defer r.Unlock()
+
+	if err != nil {
+		r.err = r.err.Add(err)
+		return
+	}
+
+	if r.result == nil {
+		r.result = result
+		r.dedupeFirstAttrs = attrs
+		return
+	}
+
+	r.result.HasNext = r.result.HasNext && result.HasNext
+	r.result.LocalOnly = r.result.LocalOnly && result.LocalOnly
+
+	// Need to dedupe
+	if r.dedupeMap == nil {
+		r.dedupeMap = make(map[string]multiFetchResultSeries, len(r.result.SeriesList))
+		for idx, s := range r.result.SeriesList {
+			r.dedupeMap[s.Name()] = multiFetchResultSeries{
+				idx:   idx,
+				attrs: r.dedupeFirstAttrs,
+			}
+		}
+	}
+
+	for _, s := range result.SeriesList {
+		id := s.Name()
+		existing, exists := r.dedupeMap[id]
+		if exists && existing.attrs.Resolution <= attrs.Resolution {
+			// Already exists and resolution is already more finer grained
```

Comment:
nit: `already finer`

Reply:
Already exists and resolution is higher ?

Reply:
Opted for `// Already exists and resolution of result we are adding is not as precise`.  The "higher" or "lower" is confusing with resolution, so I went with "precise" terminology.